File: src/rs/texture/procedural/operation/TextureOperationFactory.ts
import { ByteBuffer } from "../../../io/ByteBuffer";
import { ArithmeticOperation } from "./ArithmeticOperation";
import { BlurOperation } from "./BlurOperation";
import { BricksOperation } from "./BricksOperation";
import { BrightnessOperation } from "./BrightnessOperation";
import { ClampOperation } from "./ClampOperation";
import { ColourStripOperation } from "./ColourStripOperation";
import { ConstantColourOperation } from "./ConstantColourOperation";
import { ConstantMonochromeOperation } from "./ConstantMonochromeOperation";
import { CurveOperation } from "./CurveOperation";
import { DiagonalGradientOperation } from "./DiagonalGradientOperation";
import { EmbossOperation } from "./EmbossOperation";
import { GradientOperation } from "./GradientOperation";
import { GrayScaleOperation } from "./GrayScaleOperation";
import { HerringboneOperation } from "./HerringboneOperation";
import { HorizontalGradientOperation } from "./HorizontalGradientOperation";
import { HslOperation } from "./HslOperation";
import { InvertOperation } from "./InvertOperation";
import { IrregularBricksOperation } from "./IrregularBricksOperation";
import { KaleidoscopeOperation } from "./KaleidoscopeOperation";
import { LerpOperation } from "./LerpOperation";
import { LineNoiseOperation } from "./LineNoiseOperation";
import { MandelbrotOperation } from "./MandelbrotOperation";
import { MirrorOperation } from "./MirrorOperation";
import { MonochromeEdgeDetectorOperation } from "./MonochromeEdgeDetectorOperation";
import { NormalMapOperation } from "./NormalMapOperation";
import { PerlinNoiseOperation } from "./PerlinNoiseOperation";
import { PseudoRandomNoiseOperation } from "./PseudoRandomNoiseOperation";
import { RangeOperation } from "./RangeOperation";
import { RangeThresholdOperation } from "./RangeThresholdOperation";
import { ShapeRasterizerOperation } from "./ShapeRasterizerOperation";
import { SpriteSourceOperation } from "./SpriteSourceOperation";
import { SquareWaveformOperation } from "./SquareWaveformOperation";
import { TextureOperation } from "./TextureOperation";
import { TextureSourceOperation } from "./TextureSourceOperation";
import { TilingOperation } from "./TilingOperation";
import { TilingSpriteOperation } from "./TilingSpriteOperation";
import { TrigWarpOperation } from "./TrigWarpOperation";
import { VerticalGradientOperation } from "./VerticalGradientOperation";
import { VoronoiNoiseOperation } from "./VoronoiNoiseOperation";
import { WavyCrossOperation } from "./WavyCrossOperation";
import { WeaveOperation } from "./WeaveOperation";

export enum TextureOperationTypeId {
    ConstantMonochrome = 0,
    ConstantColour = 1,
    HorizontalGradient = 2,
    VerticalGradient = 3,
    Bricks = 4,
    Blur = 5,
    Clamp = 6,
    Arithmetic = 7,
    Curve = 8,
    Mirror = 9,
    Gradient = 10,
    ColourStrip = 11,
    DiagonalGradient = 12,
    PseudoRandomNoise = 13,
    Weave = 14,
    VoronoiNoise = 15,
    Herringbone = 16,
    Hsl = 17,
    TilingSprite = 18,
    TrigWarp = 19,
    Tiling = 20,
    Lerp = 21,
    Invert = 22,
    Kaleidoscope = 23,
    GrayScale = 24,
    Brightness = 25,
    RangeThreshold = 26,
    SquareWaveform = 27,
    IrregularBricks = 28,
    ShapeRasterizer = 29,
    Range = 30,
    Mandelbrot = 31,
    Emboss = 32,
    NormalMap = 33,
    PerlinNoise = 34,
    MonochromeEdgeDetector = 35,
    TextureSource = 36,
    WavyCross = 37,
    LineNoise = 38,
    SpriteSource = 39,
}

export class TextureOperationFactory {
    static instantiate(typeId: TextureOperationTypeId): TextureOperation {
        switch (typeId) {
            case TextureOperationTypeId.ConstantMonochrome:
                return new ConstantMonochromeOperation();
            case TextureOperationTypeId.ConstantColour:
                return new ConstantColourOperation();
            case TextureOperationTypeId.HorizontalGradient:
                return new HorizontalGradientOperation();
            case TextureOperationTypeId.VerticalGradient:
                return new VerticalGradientOperation();
            case TextureOperationTypeId.Bricks:
                return new BricksOperation();
            case TextureOperationTypeId.Blur:
                return new BlurOperation();
            case TextureOperationTypeId.Clamp:
                return new ClampOperation();
            case TextureOperationTypeId.Arithmetic:
                return new ArithmeticOperation();
            case TextureOperationTypeId.Curve:
                return new CurveOperation();
            case TextureOperationTypeId.Mirror:
                return new MirrorOperation();
            case TextureOperationTypeId.Gradient:
                return new GradientOperation();
            case TextureOperationTypeId.ColourStrip:
                return new ColourStripOperation();
            case TextureOperationTypeId.DiagonalGradient:
                return new DiagonalGradientOperation();
            case TextureOperationTypeId.PseudoRandomNoise:
                return new PseudoRandomNoiseOperation();
            case TextureOperationTypeId.Weave:
                return new WeaveOperation();
            case TextureOperationTypeId.VoronoiNoise:
                return new VoronoiNoiseOperation();
            case TextureOperationTypeId.Herringbone:
                return new HerringboneOperation();
            case TextureOperationTypeId.Hsl:
                return new HslOperation();
            case TextureOperationTypeId.TilingSprite:
                return new TilingSpriteOperation();
            case TextureOperationTypeId.TrigWarp:
                return new TrigWarpOperation();
            case TextureOperationTypeId.Tiling:
                return new TilingOperation();
            case TextureOperationTypeId.Lerp:
                return new LerpOperation();
            case TextureOperationTypeId.Invert:
                return new InvertOperation();
            case TextureOperationTypeId.Kaleidoscope:
                return new KaleidoscopeOperation();
            case TextureOperationTypeId.GrayScale:
                return new GrayScaleOperation();
            case TextureOperationTypeId.Brightness:
                return new BrightnessOperation();
            case TextureOperationTypeId.RangeThreshold:
                return new RangeThresholdOperation();
            case TextureOperationTypeId.SquareWaveform:
                return new SquareWaveformOperation();
            case TextureOperationTypeId.IrregularBricks:
                return new IrregularBricksOperation();
            case TextureOperationTypeId.ShapeRasterizer:
                return new ShapeRasterizerOperation();
            case TextureOperationTypeId.Range:
                return new RangeOperation();
            case TextureOperationTypeId.Mandelbrot:
                return new MandelbrotOperation();
            case TextureOperationTypeId.Emboss:
                return new EmbossOperation();
            case TextureOperationTypeId.NormalMap:
                return new NormalMapOperation();
            case TextureOperationTypeId.PerlinNoise:
                return new PerlinNoiseOperation();
            case TextureOperationTypeId.MonochromeEdgeDetector:
                return new MonochromeEdgeDetectorOperation();
            case TextureOperationTypeId.TextureSource:
                return new TextureSourceOperation();
            case TextureOperationTypeId.WavyCross:
                return new WavyCrossOperation();
            case TextureOperationTypeId.LineNoise:
                return new LineNoiseOperation();
            case TextureOperationTypeId.SpriteSource:
                return new SpriteSourceOperation();
            default:
                throw new Error("Unknown texture operation: " + typeId);
        }
    }

    static create(buffer: ByteBuffer): TextureOperation {
        const operationId = buffer.readUnsignedByte();
        const typeId = buffer.readUnsignedByte() as TextureOperationTypeId;
        // console.log("type", typeId, "id", operationId);
        const operation = TextureOperationFactory.instantiate(typeId);
        operation.operationId = operationId;
        operation.cacheSlotCount = buffer.readUnsignedByte();
        const propertyCount = buffer.readUnsignedByte();
        for (let i = 0; i < propertyCount; i++) {
            const propertyId = buffer.readUnsignedByte();
            operation.decode(propertyId, buffer);
        }
        operation.init();
        return operation;
    }
}

File: cpp/rs/texture/procedural/operation/TextureOperationFactory.cpp
#include "TextureOperationFactory.hpp"

#include <cstddef>
#include <new>

#include "../../../core/Move.hpp"
#include "../../../core/ScopeGuard.hpp"
#include "../../../io/Uint8ArrayReader.hpp"
#include "ArithmeticOperation.hpp"
#include "BlurOperation.hpp"
#include "BricksOperation.hpp"
#include "BrightnessOperation.hpp"
#include "ClampOperation.hpp"
#include "ColourStripOperation.hpp"
#include "ConstantColourOperation.hpp"
#include "ConstantMonochromeOperation.hpp"
#include "CurveOperation.hpp"
#include "DiagonalGradientOperation.hpp"
#include "IrregularBricksOperation.hpp"
#include "HslOperation.hpp"
#include "KaleidoscopeOperation.hpp"
#include "GradientOperation.hpp"
#include "GrayScaleOperation.hpp"
#include "HerringboneOperation.hpp"
#include "HorizontalGradientOperation.hpp"
#include "InvertOperation.hpp"
#include "LerpOperation.hpp"
#include "LineNoiseOperation.hpp"
#include "EmbossOperation.hpp"
#include "PerlinNoiseOperation.hpp"
#include "MirrorOperation.hpp"
#include "MonochromeEdgeDetectorOperation.hpp"
#include "NormalMapOperation.hpp"
#include "PseudoRandomNoiseOperation.hpp"
#include "RangeOperation.hpp"
#include "RangeThresholdOperation.hpp"
#include "ShapeRasterizerOperation.hpp"
#include "SquareWaveformOperation.hpp"
#include "SpriteSourceOperation.hpp"
#include "TilingSpriteOperation.hpp"
#include "TilingOperation.hpp"
#include "TextureSourceOperation.hpp"
#include "TrigWarpOperation.hpp"
#include "VerticalGradientOperation.hpp"
#include "VoronoiNoiseOperation.hpp"
#include "WavyCrossOperation.hpp"
#include "WeaveOperation.hpp"

namespace rs {

template <typename T>
static Result<T*> allocNew(Allocator& alloc) noexcept {
    void* mem = alloc.alloc(alloc.ctx, sizeof(T), alignof(T));
    if (!mem) {
        return Result<T*>::err(Status::OutOfMemory);
    }
    T* obj = new (mem) T();
    return Result<T*>::ok(obj);
}

static Result<TextureOperation*> instantiate(TextureOperationTypeId typeId, Allocator& alloc) noexcept {
    switch (typeId) {
    case TextureOperationTypeId::ConstantMonochrome: {
        auto r = allocNew<ConstantMonochromeOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::ConstantColour: {
        auto r = allocNew<ConstantColourOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::HorizontalGradient: {
        auto r = allocNew<HorizontalGradientOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::VerticalGradient: {
        auto r = allocNew<VerticalGradientOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::Bricks: {
        auto r = allocNew<BricksOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::Blur: {
        auto r = allocNew<BlurOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::Clamp: {
        auto r = allocNew<ClampOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::Arithmetic: {
        auto r = allocNew<ArithmeticOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::Curve: {
        auto r = allocNew<CurveOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::Gradient: {
        auto r = allocNew<GradientOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::ColourStrip: {
        auto r = allocNew<ColourStripOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::DiagonalGradient: {
        auto r = allocNew<DiagonalGradientOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::Mirror: {
        auto r = allocNew<MirrorOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::RangeThreshold: {
        auto r = allocNew<RangeThresholdOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::PseudoRandomNoise: {
        auto r = allocNew<PseudoRandomNoiseOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::Lerp: {
        auto r = allocNew<LerpOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::Invert: {
        auto r = allocNew<InvertOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::Kaleidoscope: {
        auto r = allocNew<KaleidoscopeOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::GrayScale: {
        auto r = allocNew<GrayScaleOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::Brightness: {
        auto r = allocNew<BrightnessOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::TextureSource: {
        auto r = allocNew<TextureSourceOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::SpriteSource: {
        auto r = allocNew<SpriteSourceOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::PerlinNoise: {
        auto r = allocNew<PerlinNoiseOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::MonochromeEdgeDetector: {
        auto r = allocNew<MonochromeEdgeDetectorOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::Range: {
        auto r = allocNew<RangeOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::NormalMap: {
        auto r = allocNew<NormalMapOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::WavyCross: {
        auto r = allocNew<WavyCrossOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::Weave: {
        auto r = allocNew<WeaveOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::Herringbone: {
        auto r = allocNew<HerringboneOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::IrregularBricks: {
        auto r = allocNew<IrregularBricksOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::Emboss: {
        auto r = allocNew<EmbossOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::VoronoiNoise: {
        auto r = allocNew<VoronoiNoiseOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::Hsl: {
        auto r = allocNew<HslOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::TilingSprite: {
        auto r = allocNew<TilingSpriteOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::TrigWarp: {
        auto r = allocNew<TrigWarpOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::Tiling: {
        auto r = allocNew<TilingOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::SquareWaveform: {
        auto r = allocNew<SquareWaveformOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::LineNoise: {
        auto r = allocNew<LineNoiseOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    case TextureOperationTypeId::ShapeRasterizer: {
        auto r = allocNew<ShapeRasterizerOperation>(alloc);
        if (!r.isOk()) {
            return Result<TextureOperation*>::err(r.status());
        }
        return Result<TextureOperation*>::ok(r.value());
    }
    default:
        return Result<TextureOperation*>::err(Status::Unsupported);
    }
}

Result<TextureOperationCreateResult> createTextureOperation(Uint8ArrayReader& reader, Allocator& alloc) noexcept {
    u8 opId = 0;
    u8 typeIdU8 = 0;
    u8 cacheSlotCount = 0;
    u8 propertyCount = 0;

    Status s = reader.readUnsignedByte(&opId);
    if (!ok(s)) {
        return Result<TextureOperationCreateResult>::err(s);
    }
    s = reader.readUnsignedByte(&typeIdU8);
    if (!ok(s)) {
        return Result<TextureOperationCreateResult>::err(s);
    }
    s = reader.readUnsignedByte(&cacheSlotCount);
    if (!ok(s)) {
        return Result<TextureOperationCreateResult>::err(s);
    }
    s = reader.readUnsignedByte(&propertyCount);
    if (!ok(s)) {
        return Result<TextureOperationCreateResult>::err(s);
    }

    auto opRes = instantiate(static_cast<TextureOperationTypeId>(typeIdU8), alloc);
    if (!opRes.isOk()) {
        return Result<TextureOperationCreateResult>::err(opRes.status());
    }
    TextureOperation* op = opRes.value();

    auto opGuard = makeScopeGuard([&]() noexcept {
        if (op) {
            op->destroy(alloc);
        }
    });

    op->setHeader(static_cast<i32>(opId), cacheSlotCount);

    for (u8 i = 0; i < propertyCount; i++) {
        u8 fieldId = 0;
        s = reader.readUnsignedByte(&fieldId);
        if (!ok(s)) {
            return Result<TextureOperationCreateResult>::err(s);
        }
        s = op->decode(fieldId, reader, alloc);
        if (!ok(s)) {
            return Result<TextureOperationCreateResult>::err(s);
        }
    }
    s = op->init();
    if (!ok(s)) {
        return Result<TextureOperationCreateResult>::err(s);
    }

    TextureOperationCreateResult out{};
    out.op = op;
    out.inputCount = op->inputCount();
    opGuard.dismiss();
    return Result<TextureOperationCreateResult>::ok(out);
}

} // namespace rs
